### DISCUSSION TITLE
How to add URL variable

### DISCUSSION THREAD

**Discussion by brandon-claps** [mshibanami/redirect-web#39] [Redirect Rules]
Hello,

I want to add a URL variable on to an existing URL by adding ? at the end of the URL. If there are existing URL variables I need to add it by using &. Can you help with the appropriate rules?

**Comment by mshibanami** [comment]
Hi @S10XtremeNLow. If you want to handle a URL that does/doesn't include existing variables (query parameters), you can use a pipe (`|`) of Regular Expression.

For example, suppose you want to add `?newkey=value` to `https://example.com/page` or `https://example.com/page?existingkey=value`, the rule would look like this:

- Redirect From: `(https://example.com/page)(?:\?(.*)|)` (Regular Expression)
- Redirect To: `$1\?newkey=value&$2`
- Excluded URL pattern `.*[&?]newkey=value.*`
    - This is to avoid an infinite loop

![Screenshot 2024-06-17 at 9 52 07 PM](https://github.com/mshibanami/redirect-web/assets/1333214/eed6d439-3ce4-466c-9e46-6ff0b005c52e)

Note: as you can see at the end of the above screenshot, an unnecessary `&` is appended at the end of the URL (= `https://example.com/page?newkey=value&`).

---

Or, you can create a separate rule so that each rule handles either.

**Reply by brandon-claps** [reply]
Hi @mshibanami  

thanks for the quick reply. This took care of the infinite loop issue I was having but looks like it may not be working for me. Here is an example of what the URL could look like.

https://www.test.com/02/history/test.html

https://test.com/test.html

https://www.test.com/2020/05/06/03/test.html?utmsource=blah

**Reply by mshibanami** [reply]
@S10XtremeNLow I see. Then, maybe like this? 🙂 

- Redirect From: `(https://(?:www\.)?test.com/[^?]*)(?:\?(.*)|)` (Regular Expression)
- Redirect To: `$1\?newkey=value&$2`
- Excluded URL Pattern `.*[&?]newkey=value.*` (Regular Expression)

![Screenshot 2024-06-17 at 10 41 45 PM](https://github.com/mshibanami/redirect-web/assets/1333214/3e6ad24d-e9d4-4bfa-8bce-b75596dcf97b)
